In [18]:
# Initialize Otter
import otter
grader = otter.Notebook("hwk1-task2-corals.ipynb")

FileNotFoundError: [Errno 2] No such file or directory: 'hwk1-task2-corals.ipynb'

In [3]:
# Initialize Otter
import otter
grader = otter.Notebook("hwk1-task2-corals-WITT.ipynb")

# Task 2: Exploring coral diversity data

## Instructions 

- First, update the following cell to have a link to *your* Homework 1 GitHub repository:


**UPDATE THIS LINK:**

https://github.com/olivrwitt/eds220-hwk1.git

- This task covers topics on the lesson on [subsetting `pandas.DataFrames`](https://meds-eds-220.github.io/MEDS-eds-220-course/book/chapters/lesson-3-pandas-subsetting/lesson-3-pandas-subsetting.html). The best strategy to solve it is to get together with your classmates and work on it together after that lecture.

- Review the [complete rubric for this task](https://docs.google.com/document/d/131OnUDOLpOl94SlaMU9nGqT998ABrjFS/edit?usp=sharing&ouid=111507336322611936333&rtpof=true&sd=true) before starting.

- **Make at least 5 commits at moments you think are adequate.** We'll check your repository and view the commit history.

- **Add comments for all your code.** Err on the side of commenting too much for now. Comments should follow best practices.



<!-- BEGIN QUESTION -->


## 1
For this task we are going to use data about  [Western Indian Ocean Coral Diversity (McClanahan, 2023)](https://knb.ecoinformatics.org/view/doi%3A10.5063%2FF1K35S3H) from the Knowledge Network for Biocomplexity (KNB) data repository. 

a. Navigate to the dataset's link and explore the data package. 
 
b. Use the next markdown cell to add a brief description of the dataset, including a citation, date of access, and a link to the archive. 


Dataset title: Western Indian Ocean Coral Diversity

Authors: Tim McClanahan

Date of publication: 2023

Date of access: Oct 8, 2026

Link/URL: https://knb.ecoinformatics.org/view/doi%3A10.5063%2FF1K35S3H\

Citation: Tim McClanahan. (2023). Western Indian Ocean Coral Diversity. Knowledge Network for Biocomplexity. doi:10.5063/F1K35S3H.

Contents: This dataset contains data on coral biodiversity in the Indian Ocean in the form of coral coverage and number of genera.  Data was collected via SCUBA or snorkel surveys in a variety of countries in the Wester Indian Ocean region.  In addition to coverage and number of genera, this dataset also contains information on sample location (country, ecoregion, depth), water chemistry (dissolved oxygen, calcite concentration, salinity), sea surfaced temperature, and a climate stress model

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

## 2
In the next code cell, import the `pandas` package using standard abbreviation in a code cell.

Read in the data from the URL using the `pd.read_csv` function and store it as `coral_div` like this:

```
# read in data
coral_div = pd.read_csv('the URL goes here')
```

In [6]:
import pandas as pd
coral_div = pd.read_csv('data/coral_data.csv')

<!-- END QUESTION -->

## 3

Use attributes and methods of the `coral_div` data frame to obtain the following information about it:

Use the code cells marked a-e. For each, add a comment explaining what is the ouput.

a. how many rows and columns does it have? Store your answers in the `rows` and `columns` variables. 

In [9]:
#a
rows = len(coral_div)

columns = len(coral_div.columns)

print(rows, columns)
# output is (number of rows, number of columns)

575 42


In [10]:
grader.check("q3_a")

q3_a results: All test cases passed!

b. What are the column names? Store your answer in the `coral_div_columns` variable.

In [11]:
#b
coral_div_columns = coral_div.columns

In [12]:
grader.check("q3_b")

q3_b results: All test cases passed!

c. what are the data types of the columns? Store your answer in the `coral_div_types`.

In [15]:
#c
coral_div_types = coral_div.dtypes
# stores the data types of the information in each column

In [16]:
grader.check("q3_c")

q3_c results: All test cases passed!

d. display the first 8 rows with all the column names. For this one, you will need to udpate a `pandas` option (HINT: check the first discussion section). Store your answer in the `first_8_rows` variable.

In [20]:
#d
first_8_rows = coral_div.head(8)
# head displays the first few rows and the number in parantheses specifies how many rows

In [18]:
# View the dataframe 
first_8_rows

,PU_ID,PU_Lat,PU_Lon,Number_of_genera,Country,Ecoregion,Management,Habitat,Observer,Depth,...,SST_bimodality,SST_Skewness,ClimateStressModel,Netflow,Indegree,Outdegree,Retention,min_yr,max_yr,no_replicates
0,7,39.076711,-5.490011,35,Tanzania,East African Coral Coast,UnfishedLow,Crest,Tim,11.0,...,0.462556,-0.128136,0.596375,0.144498,98.25,74.75,14.00,2019,2019,1
1,7,39.076711,-5.490011,22,Tanzania,East African Coral Coast,UnfishedLow,Lagoon/Back reef,Nyawira,2.0,...,0.462556,-0.128136,0.596375,0.144498,98.25,74.75,14.00,2010,2015,2
2,7,39.076711,-5.490011,25,Tanzania,East African Coral Coast,UnfishedLow,Slope,Nyawira,4.0,...,0.462556,-0.128136,0.596375,0.144498,98.25,74.75,14.00,2015,2015,1
3,7,39.076711,-5.490011,26,Tanzania,East African Coral Coast,UnfishedLow,Slope,Nyawira,3.0,...,0.462556,-0.128136,0.596375,0.144498,98.25,74.75,14.00,2016,2019,2
4,10,39.054253,-5.512514,27,Tanzania,East African Coral Coast,UnfishedLow,Lagoon/Back reef,Nyawira,1.5,...,0.461297,-0.134548,0.629000,0.011721,112.75,51.50,183.75,2010,2019,3
5,10,39.054253,-5.512514,29,Tanzania,East African Coral Coast,UnfishedLow,Lagoon/Back reef,Tim,1.5,...,0.461297,-0.134548,0.629000,0.011721,112.75,51.50,183.75,2015,2015,1
6,15,39.054253,-5.535017,19,Tanzania,East African Coral Coast,Restricted,Crest,Tim,8.0,...,0.461297,-0.134548,0.595600,0.011721,112.75,51.50,183.75,2019,2019,1
7,39,39.009337,-5.647518,28,Tanzania,East African Coral Coast,UnfishedLow,Lagoon/Back reef,Nyawira,3.0,...,0.464874,-0.158601,0.626000,-0.063087,96.50,40.75,35.75,2015,2019,3


In [19]:
grader.check("q3_d")

q3_d results: All test cases passed!

<!-- BEGIN QUESTION -->

e. Write code to find out some other inital information you are interested in. 

In [28]:
# e
print(coral_div["Country"].nunique())
# shows how many different countries were sampled
print(coral_div["Country"].unique())
# prints the names of each country that was sampled

# I am also curious about how many ecoregions were sampled:
print(coral_div["Ecoregion"].nunique())
print(coral_div["Ecoregion"].unique())

11
['Tanzania' 'Mozambique' 'Kenya' 'Reunion' 'Comoros' 'Mayotte'
 'French Southern Territories' 'Madagascar' 'Seychelles' 'South Africa'
 'Mauritius']
6
['East African Coral Coast' 'Delagoa' 'Northern Monsoon Current Coast'
 'Mascarene Islands' 'Western and Northern Madagascar' 'Seychelles']


<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

## 4

Some of the columns have data type `int64` and some have data type `float64`. What is the difference between these `int64` and `float64`?


int64 signifies that the data type in that column is an integer (i.e. 1,2,3,etc.).  float64 signifies that the data type in that column is a float (i.e. a decimal number like 1.75, 10,076.99, etc.)

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

## 5

The `Country` column indicates the country where the coral reef is located. Explain, in your own words, what is the ouput of `coral_div['Country']` and `coral_div['Country'].unique()`. 

The output of coral_div['Country'] will be every entry in the country column in the coral_div dataset (all 575 country observations).  The output of coral_div['Country'].unique() will just be each unique country observation.  In this case, the 575 country observations consisted of 11 unique countries, so the output of coral_div['Country'].unique() will be a list of those 11 countries. 

<!-- END QUESTION -->

## 6

Check the dataset's [metadata file README_md.doc](https://knb.ecoinformatics.org/view/doi:10.5063/F1K35S3H) to find which column has information about who made an observation. How many observers collected this data? HINT: Check this [method for `pandas.DataFrame`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.nunique.html). Store your answer in the `unique_observers` variable.

In [41]:
unique_observers = coral_div["Observer"].nunique()
unique_observers

16

In [42]:
grader.check("q6")

q6 results: All test cases passed!

## 7

The `Coralcover` column indicates the percent of hard coral cover in each 6 km reef cell surveyed. 

<!-- BEGIN QUESTION -->

a)  Describe, in detail,  what kind of object `coral_div['Coralcover']>= 50` is.

coral_div['Coralcover']>= 50 produces a true/false (boolean) list, with 575 entries (one per observation) signifying whether or not the coral cover observed in that observation was greater than 50% (true = observed coral cover was greater than or equal to 50%, and false = observed coral cover was not greater than or equal to 50%).

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

b) What data is `coral_div[coral_div['Coralcover'] >= 50]` selecting?

This code selects only those columns in the coral_div dataset where the observed coral cover was greater than or equal to 50%

<!-- END QUESTION -->

## 8

Use code to find if there are there any sites with a 30% coral cover in Tanzania. Store your answer in the `coral_cover` variable.

In [43]:
tanzania_coral = coral_div[coral_div["Country"] == "Tanzania"]
coral_cover = tanzania_coral[tanzania_coral["Coralcover"] == 30]

In [44]:
# View the dataframe
coral_cover
# 9 observations of 30% coral cover in Tanzania

,PU_ID,PU_Lat,PU_Lon,Number_of_genera,Country,Ecoregion,Management,Habitat,Observer,Depth,...,SST_bimodality,SST_Skewness,ClimateStressModel,Netflow,Indegree,Outdegree,Retention,min_yr,max_yr,no_replicates
63,693,39.391121,-5.782490,10,Tanzania,East African Coral Coast,Fished,Lagoon/Back reef,Nyawira,5.000000,...,0.490512,-0.101724,0.655600,-0.309514,77.00,65.25,147.50,2018,2018,2
64,697,39.391121,-5.804983,22,Tanzania,East African Coral Coast,Fished,Lagoon/Back reef,Nyawira,9.166667,...,0.490769,-0.100365,0.683050,-0.309514,77.00,65.25,147.50,2010,2018,3
341,4294,39.211458,-4.814515,37,Tanzania,East African Coral Coast,Fished,Slope,Nyawira,11.000000,...,0.451293,-0.048992,0.608900,0.514920,46.00,49.50,82.75,2018,2018,1
342,4299,39.233916,-4.837043,26,Tanzania,East African Coral Coast,Fished,Slope,Nyawira,12.500000,...,0.451293,-0.048992,0.601725,0.514920,46.00,49.50,82.75,2018,2018,2
343,4302,39.211458,-4.859570,31,Tanzania,East African Coral Coast,Fished,Slope,Nyawira,7.000000,...,0.452402,-0.050477,0.595200,0.514920,46.00,49.50,82.75,2018,2018,1
344,4303,39.301290,-4.859570,23,Tanzania,East African Coral Coast,Fished,Slope,Nyawira,11.000000,...,0.453806,-0.048007,0.602300,0.514920,46.00,49.50,82.75,2018,2018,1
377,4906,39.660616,-4.972195,33,Tanzania,East African Coral Coast,Fished,Lagoon/Back reef,Nyawira,9.000000,...,0.488396,-0.074229,0.671900,0.156359,66.75,41.75,38.00,2013,2013,2
388,5049,39.615700,-5.490011,14,Tanzania,East African Coral Coast,Fished,Lagoon/Back reef,Nyawira,4.000000,...,0.489921,-0.109523,0.657000,-0.328399,93.25,57.25,650.25,2018,2018,1
389,5050,39.638158,-5.490011,18,Tanzania,East African Coral Coast,Fished,Lagoon/Back reef,Nyawira,6.000000,...,0.489921,-0.109523,0.618825,-0.328399,93.25,57.25,650.25,2018,2018,1


In [45]:
grader.check("q8")

q8 results: All test cases passed!

## 9

We are interested in looking at the PH of sea water (`PH` column),  mean salinity (`Salinity_mean` column), and depth of survey location in meters (`Depth` column). Subset the data frame by selecting only these column. Store your answer in the `filtered_data` variable

In [46]:
filtered_data = coral_div[["PH", "Salinity_mean", "Depth"]]

In [47]:
# View the dataframe
filtered_data.head()

,PH,Salinity_mean,Depth
0,8.2175,35.167646,11.0
1,8.2175,35.167646,2.0
2,8.2175,35.167646,4.0
3,8.2175,35.167646,3.0
4,8.2180,35.102688,1.5


In [48]:
grader.check("q9")

q9 results: All test cases passed!

## 10
The `max_yr` column indicates the year of last sampling at a given site. Select the PH, salinity, and depth for sites where sampling ended in 2017. Store your answer in the `sampling_2017` variable. `Sampling_2017` should include the following columns in the following order: ` PH, Salinity_mean, Depth, max_yr`. 


In [53]:
yr_is_2017 = coral_div[coral_div["max_yr"] == 2017]
sampling_2017 = yr_is_2017[["PH", "Salinity_mean", "Depth", "max_yr"]]

In [54]:
# View the dataframe 
sampling_2017

,PH,Salinity_mean,Depth,max_yr
67,8.2150,35.100085,16.000000,2017
68,8.2145,35.144927,10.000000,2017
70,8.2150,35.101086,11.000000,2017
71,8.2150,35.101086,11.000000,2017
72,8.2150,35.101086,9.000000,2017
73,8.2150,35.101086,9.000000,2017
84,8.2110,35.278084,12.500000,2017
85,8.2110,35.278084,6.000000,2017
164,8.2220,35.337914,9.875000,2017
167,8.2220,35.337914,12.075000,2017


In [55]:
grader.check("q10")

q10 results: All test cases passed!

## 11
Use `iloc` to select rows 95 through 100 (including 100) in the 17th, 20th, and 28th columns and assign these values to a new variable called `subset`. View the new `subset` dataframe.

HINT: the ouput columns should be `Salinity_mean`, `mean.npp` and `Current_vel_mean` (in that order!).

In [64]:
subset = coral_div.iloc[95:101, [16,19,27]]

In [65]:
# View the dataframe
subset

,Salinity_mean,mean.npp,Current_vel_mean
95,35.201630,1365.146729,0.878417
96,35.204185,1117.635010,0.839372
97,35.234774,1105.270142,0.685955
98,35.234774,1105.270142,0.685955
99,35.250068,1082.196411,NaN
100,35.250068,1082.196411,NaN


In [66]:
grader.check("q11")

q11 results: All test cases passed!

<!-- BEGIN QUESTION -->

## 12 

Run `subset.isna()`. Explain what this code does and what the output represents.

In [67]:
subset.isna()

,Salinity_mean,mean.npp,Current_vel_mean
95,False,False,False
96,False,False,False
97,False,False,False
98,False,False,False
99,False,False,True
100,False,False,True


This code looks at each cell in the subset dataframe that we previously created (rows 95-100 and the salinity, mean npp, and current velocity mean columns).  If that cell contains an NA value, subset.isna() returns true, otherwise it returns false.

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

## 13
Run `subset.isna().any()`. Explain in your own words what this code does and what the output represents. HINT: check the [documentation of `pandas.Sries.any`](https://pandas.pydata.org/docs/reference/api/pandas.Series.any.html).


In [68]:
subset.isna().any()

Salinity_mean       False
mean.npp            False
Current_vel_mean     True
dtype: bool

This code looks at each column in the previously created subset and returns true if there are any NA values in that column; otherwise it returns false.  The output is a boolean series reporting whether there are any NAs.

<!-- END QUESTION -->

## 14 

Select the rows in `subset` that have NA in the `Current_vel_mean` column. Store your answer in the `subset_na` variable.

In [70]:
subset_na = subset[subset["Current_vel_mean"].isna() == True]

In [71]:
# View the dataframe 
subset_na

,Salinity_mean,mean.npp,Current_vel_mean
99,35.250068,1082.196411,NaN
100,35.250068,1082.196411,NaN


In [72]:
grader.check("q14")

q14 results: All test cases passed!

---

To double-check your work, the cell below will rerun all of the autograder tests.

In [73]:
grader.check_all()

q10 results: All test cases passed!

q11 results: All test cases passed!

q14 results: All test cases passed!

q3_a results: All test cases passed!

q3_b results: All test cases passed!

q3_c results: All test cases passed!

q3_d results: All test cases passed!

q6 results: All test cases passed!

q8 results: All test cases passed!

q9 results: All test cases passed!